Sources/MongoSwift/Operations/ListDatabasesOperation.swift

```diff
@@ -0,0 +1,91 @@
+    /// This is internal and only used for implementation purposes. Users should use `listDatabaseNames` if they want
+    /// only names.
+    internal var nameOnly: Bool?
+
+    /// Convenience constructor for basic construction
+    internal init(filter: Document?, nameOnly: Bool?) {
+        self.filter = filter
+        self.nameOnly = nameOnly
+    }
+}
+
+/// A struct modeling the information returned from the `listDatabases` command about a single database.
+public struct DatabaseSpecification: Codable {
+    /// The name of the database.
+    public let name: String
+
+    /// The amount of disk space consumed by this database.
+    public let sizeOnDisk: Int
+
+    /// Whether or not this database is empty.
+    public let empty: Bool
+
+    /// For sharded clusters, this field includes a document which maps each shard to the size in bytes of the database
+    /// on disk on that shard. For non sharded environments, this field is nil.
+    public let shards: Document?
+}
+
+/// Internal intermediate result of a ListDatabases command.
+internal enum ListDatabasesResults {
+    /// Includes the names and sizes.
+    case specs([DatabaseSpecification])
+
+    /// Only includes the names.
+    case names([String])
+}
+
+/// An operation corresponding to a "listDatabases" command on a collection.
+internal struct ListDatabasesOperation {
+    private let session: ClientSession?
+    private let client: MongoClient
+    private let options: ListDatabasesOptions?
+
+    internal init(client: MongoClient,
+                  options: ListDatabasesOptions?,
+                  session: ClientSession?) {
+        self.client = client
+        self.options = options
+        self.session = session
+    }
+
+    internal func execute() throws -> ListDatabasesResults {
+        // spec requires that this command be run against the primary.
+        let readPref = ReadPreference(.primary)
+        let cmd: Document = ["listDatabases": 1]
+        let opts = try encodeOptions(options: self.options, session: self.session)
+        var reply = Document()
+        var error = bson_error_t()
+
+        let success = withMutableBSONPointer(to: &reply) { replyPtr in
+            mongoc_client_read_command_with_opts(self.client._client,
+                                                 "admin",
+                                                 cmd._bson,
+                                                 readPref._readPreference,
+                                                 opts?._bson,
+                                                 replyPtr,
+                                                 &error)
+        }
+
+        guard success else {
+            throw extractMongoError(error: error, reply: reply)
+        }
+
+        guard let databases = reply["databases"] as? [Document] else {
+            throw RuntimeError.internalError(message: "Invalid server response: \(reply)")
+        }
+
+        if self.options?.nameOnly ?? false {
```

Comment:
So yeah, this is a little weird but the rationale is as follows:
- `listDatabases` and `listDatabase` with `nameOnly` have significantly different performance characteristics due to db level locking in the former
- For that reason, we should probably provide two different APIs to communicate this (e.g. include `listDatabaseNames`)
    - Other drivers (e.g. go, libmongoc) provide separate helpers, so we get cross-driver consistency which is always a plus imo
- If we're going to provide a `listDatabaseNames`, then there is no reason a user should ever use the full `listDatabases` and provide a `nameOnly` option, so let's not allow them to.

Not allowing `nameOnly` also has the added benefit of allowing all fields on `DatabaseSpecification` to be non-optional.

Reply:
@patrickfreed we discussed this only briefly in person, but remember that the `nameOnly` is currently being serialized into the `with_opts` options, but should instead be in the direct command document above. Did you decide against that, or did it just get lost in the mix?

Reply:
Yep, totally forgot about that, good point. I've actually just gone ahead and removed the options struct since it isn't part of the public API anyways and the values in it aren't even being sent as options.